src/workerd/io/worker-fs.c++

```diff
@@ -0,0 +1,935 @@
+  if (actualSize > 0) data.asPtr().fill(0);
+  auto owned = data.attach(js.getExternalMemoryAdjustment(actualSize));
+  return kj::rc<FileImpl>(kj::mv(owned));
+}
+
+kj::Rc<File> File::newReadable(kj::ArrayPtr<const kj::byte> data) {
+  return kj::rc<FileImpl>(data);
+}
+
+kj::Own<VirtualFileSystem> newVirtualFileSystem(kj::Rc<Directory>&& root) {
+  return kj::heap<VirtualFileSystemImpl>(kj::mv(root));
+}
+
+kj::Own<VirtualFileSystem> newWorkerFileSystem(
+    const FsMap& fsMap, kj::Rc<Directory> bundleDirectory) {
+  // Our root directory is a read-only directory
+  Directory::Builder builder;
+  builder.addPath(fsMap.getBundlePath(), kj::mv(bundleDirectory));
+  builder.addPath(fsMap.getTempPath(), getTmpDirectoryImpl());
+  return newVirtualFileSystem(builder.finish());
+}
+
+kj::Rc<Directory> getTmpDirectoryImpl() {
+  return kj::rc<TmpDirectory>();
+}
+
+bool TmpDirStoreScope::hasCurrent() {
+  return tmpDirStorageScope != nullptr;
+}
+
+TmpDirStoreScope& TmpDirStoreScope::current() {
+  KJ_ASSERT(hasCurrent(), "no current TmpDirStoreScope");
+  return *static_cast<TmpDirStoreScope*>(tmpDirStorageScope);
+}
+
+TmpDirStoreScope::TmpDirStoreScope(kj::Maybe<kj::Badge<TmpDirStoreScope>> guard)
+    : dir(Directory::newWritable()) {
+  if (guard == kj::none) {
+    kj::requireOnStack(this, "must be created on the stack");
+    onStack = true;
+    KJ_ASSERT(!hasCurrent(), "TmpDirStoreScope already exists on this thread");
+    tmpDirStorageScope = this;
+  }
+}
+
+TmpDirStoreScope::~TmpDirStoreScope() noexcept(false) {
+  if (onStack) {
+    KJ_ASSERT(tmpDirStorageScope == this, "this TmpDirStoreScope not on the stack");
+    tmpDirStorageScope = nullptr;
+  }
+}
+
+kj::Own<TmpDirStoreScope> TmpDirStoreScope::create() {
+  // Creating the instance with the badge will ensure that
+  // it is not set as current in the stack.
+  return kj::heap<TmpDirStoreScope>(kj::Badge<TmpDirStoreScope>());
+}
+
+Stat SymbolicLink::stat(jsg::Lock& js) {
+  return Stat{
+    .type = FsType::SYMLINK,
+    .size = 0,
+    .lastModified = kj::UNIX_EPOCH,
+    .writable = false,
+  };
+}
+
+jsg::Url SymbolicLink::getTargetUrl() const {
+  auto path = getTargetPath().toString(false);
+  return KJ_ASSERT_NONNULL(jsg::Url::tryParse(path, "file:///"_kj));
+}
+
+kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>>> SymbolicLink::resolve(jsg::Lock& js) {
+  return root->tryOpen(js, getTargetPath());
+}
+
+kj::Rc<Directory> getLazyDirectoryImpl(kj::Function<kj::Rc<Directory>()> func) {
+  return kj::rc<LazyDirectory>(kj::mv(func));
+}
+
+kj::Maybe<const VirtualFileSystem&> tryGetCurrent(jsg::Lock&) {
+  // Note that the jsg::Lock& argument here is not actually used. We require
+  // that a jsg::Lock reference is passed in as proof that current() is called
+  // from within a valid isolate lock so that the Worker::Api::current()
+  // call below will work as expected.
+  return Worker::Api::current().getVirtualFileSystem();
```

Comment:
Bah, this won't work in python 😢 
`Api::current` uses `Worker::Isolate::Impl::Lock` which is not instantiated when we do python startups

Reply:
That's part of why I suggested we might have to go with the abstract base class that is passed into jsg